### DISCUSSION TITLE
Architecture & Idea & Design Pattern

### DISCUSSION THREAD

**Discussion by YeomanYe** [eclipse-theia/theia#10041] [Q&A]
Thank you very much for providing this tool, it is very useful for me.So I want to learn about this impressive and exciting open source technology.But when I looked up materials and read codes, there were a lot of things I couldn't understand.

Does Eclipse Theia use well-known architectural solutions (e.g., clean architecture, hexagonal architecture, MVC, etc.), ideas (e.g., DDD), design patterns (e.g., mediator pattern, strategy pattern)? 

What parts of the code, if any, are used?

**Comment by msujew** [comment]
Hi @YeomanYe, thank you for your question.

Theia doesn't explicitely follow an architectural pattern, as it not your usual web-app. To explain some major differences from normal apps:

* A Theia server instance is designed to run for a single client
* There's effectively no persistance (no database)
* Theia explicitely tries to be a framework for IDEs, not a fully fledged application in itself

 Especially the last point is important, as most design decisions in Theia were based around the "framework" assumption:

* We make heavy use of Dependency Injection
* Everything is overridable, we use the template method and factory patterns a lot throughout the code
* Every part (code editor, file explorer, git, etc) of the application is its own isolated package
* Everything is build on top of a "core" package, which handles the most important aspects of Theia:
    * Basic application layout
    * Menus, Commands, Trees, etc.
    * Messaging (frontend/backend communication via JSON-RPC over websockets)

Also, it supports the vscode extension protocol, which also influences design decisions a fair bit.

Hope that clarifies the basic architecture of Theia a bit.
